### PR TITLE
#15140: Fix UAF error when MeshDevice.close_devices() not invoked

### PR DESCRIPTION
**Pull request by cfjchu**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/15140)

### Problem description
When `MeshDevice::close_devices()` not invoked before program termination, there is a UAF issue with SystemMesh. 

### What's changed
1. The primary fix is changing the held ownership of MeshDevice handles in SystemMesh to be a weak reference to avoid double delete. 
2. Running ASAN build exposed an issue in `device.cpp` which is now fixed. Loop iterations didn't properly decrement `connections_remaining`
3. Added new unit test file in TT-NN.

### Checklist
- [x] Post commit CI passes:
    - https://github.com/tenstorrent/tt-metal/actions/runs/11929924625
    - https://github.com/tenstorrent/tt-metal/actions/runs/11929927118
- [n/a] Blackhole Post commit (if applicable)
- [n/a] Model regression CI testing passes (if applicable)
- [n/a] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/impl/device/device.cpp [hunk 1]
```diff
@@ -968,6 +968,7 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
                                     dispatch_constants::DISPATCH_BUFFER_LOG_PAGE_SIZE,
                                     settings.producer_semaphore_id,  // upstream sem
                                     mux_sem++); // local sem
+                        connections_remaining--;
```
Comment (cfjchu) [inline tt_metal/impl/device/device.cpp:971]: @ubcheema could you please review this line of code?

**Review by ayerofieiev-tt — COMMENTED** [review]
don't know about the change in `tt_metal/impl/device/device.cpp`, the rest looks ok

**Review by ayerofieiev-tt — APPROVED** [review]

File: tests/ttnn/distributed/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,35 @@
+add_executable(test_distributed test_distributed.cpp)
+
+# Common function to set up target properties
+function(setup_ttnn_test_target target_name)
```
Comment (tt-rkim) [inline tests/ttnn/distributed/CMakeLists.txt:4]: Copy pasta?

**Review by ubcheema — APPROVED** [review]

File: tests/ttnn/distributed/CMakeLists.txt (hunk 2 above)
Comment (cfjchu) [reply tests/ttnn/distributed/CMakeLists.txt]: I've moved this up and commonized it in the parent CMakeLists. Thanks!

File: tests/ttnn/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Common function to set up target properties for TTNN tests
+function(setup_ttnn_test_target target_name)
+    target_link_libraries(
+        ${target_name}
+        PUBLIC
+            test_common_libs
+            ttnn
+            Metalium::Metal
+            GTest::gtest
```
Comment (afuller-TT) [inline tests/ttnn/CMakeLists.txt:9]:
```suggestion
```

This one is redundant.  We're using the gtest_main interface to GTest.

**Review by afuller-TT — APPROVED** [review]
CMake files look fine.  We'll be sweeping through for cleanup later, so no biggie in this PR.

**Review by tt-asaigal — APPROVED** [review]

**Pull request merged**